tusd: init at 1.8.0

nh2 opened the pull request:
###### Description of changes

* Packages `tusd`, the reference server implementation of https://tus.io
* As requested on https://github.com/tus/tusd/issues/683
* Also a NixOS VM test
* The `curl`-based test upload script is from https://github.com/tus/tus.io/issues/96#issuecomment-312036573

I would like to have co-maintainers for this, since I'm just trying out `tusd` currently.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

nh2 committed:
tusd: init at 2.8.0

onny commented:
Result of `nixpkgs-review pr 167076` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tusd</li>
  </ul>
</details>

onny commented:
@ofborg build nixosTests.tusd

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/packages-looking-for-a-maintainer/5442/7

nh2 commented:
I converted to a draft PR because 

> I would like to have co-maintainers for this, since I'm just trying out `tusd` currently.

kalbasit commented on pkgs/servers/tusd/default.nix:
```diff
@@ -0,0 +1,24 @@
+    description = "Reference server implementation in Go of tus: the open protocol for resumable file uploads";
+    license = licenses.mit;
+    homepage = "https://tus.io/";
+    maintainers = with maintainers; [ nh2 ];
```
```suggestion
    maintainers = with maintainers; [ nh2 kalbasit ];
```

nh2 replied:
Done.

kalbasit reviewed (approved):
Shame to see your work going to waste, I volunteer to help you maintain this. Can you get it rebased? Also, you might as well move the package to the new per name format.

nh2 commented:
@kalbasit I have rebased and updated to tusd `2.8.0`. I also fixed a race condition in the test (waiting for `tusd` to listen on its port after it has started). Please take another look.

nh2 commented:
Also note that I'm not using `tusd` at all myself, as I implemented the TUS protocol directly into our server (in a way that I consider better than `tusd` because it's atomic without locks when saving to the local file system).

But I still think it makes sense to have the reference implementation in nixpkgs.

onny commented on nixos/tests/all-tests.nix:
```diff
@@ -1389,6 +1389,7 @@ in
   tuptime = handleTest ./tuptime.nix { };
   turbovnc-headless-server = handleTest ./turbovnc-headless-server.nix { };
   turn-rs = handleTest ./turn-rs.nix { };
+  tusd = handleTest ./tusd.nix {};
```
please use `runTest`

nh2 replied:
Done.

But where can I read about `runTest` vs `handleTest`?

onny commented on pkgs/by-name/tu/tusd/package.nix:
```diff
@@ -0,0 +1,30 @@
+{ lib, buildGoModule, fetchFromGitHub, nixosTests }:
```
please format all files with `nixfmt`

nh2 replied:
Done.

onny commented on pkgs/by-name/tu/tusd/package.nix:
```diff
@@ -0,0 +1,30 @@
+  src = fetchFromGitHub {
+    owner = "tus";
+    repo = "tusd";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

nh2 replied:
Done.

onny commented on pkgs/by-name/tu/tusd/package.nix:
```diff
@@ -0,0 +1,30 @@
+
+  meta = with lib; {
+    description = "Reference server implementation in Go of tus: the open protocol for resumable file uploads";
+    license = licenses.mit;
```
Please remove `with lib;` and use `lib.licenses.mit and `with lib.maintainers;`

nh2 replied:
Done.

onny commented on nixos/tests/tus-curl-upload.sh:
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
```
Would be better to have this script inside the test file

nh2 replied:
I disagree: Then I can't easily use it for local/interactive testing when it doesn't work.

onny replied:
having the file tus-curl-upload.sh just in nixos/test/ seems messy, maybe create a sub directory nixos/tests/tusd and put everything in there?

nh2 replied:
Ah, I see, yes. Done, I did that move.

nh2 commented:
@onny Thanks! I addressed the feedback, please take another look.

onny commented:
just one remark, otherwise LGTM

nh2 commented:
Done with the latest force-push, we should merge when CI is green.
